dd-sdk-android/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/SessionReplayRequestFactory.kt

```diff
@@ -6,41 +6,145 @@
+        )
+    }
+
+    private fun tags(datadogContext: DatadogContext): String {
+        val elements = mutableListOf(
+            "${RumAttributes.SERVICE_NAME}:${datadogContext.service}",
+            "${RumAttributes.APPLICATION_VERSION}:" +
+                datadogContext.version,
+            "${RumAttributes.SDK_VERSION}:${datadogContext.sdkVersion}",
+            "${RumAttributes.ENV}:${datadogContext.env}"
+        )
+        if (datadogContext.variant.isNotEmpty()) {
+            elements.add("${RumAttributes.VARIANT}:${datadogContext.variant}")
+        }
+        return elements.joinToString(",")
+    }
+
+    private fun buildUrl(datadogContext: DatadogContext): String {
+        val queryParams = buildQueryParameters(datadogContext)
+        return intakeUrl + queryParams.map { "${it.key}=${it.value}" }
+            .joinToString("&", prefix = "?")
+    }
+
+    private fun resolveHeaders(datadogContext: DatadogContext, requestId: String):
+        Map<String, String> {
+        return mapOf(
+            RequestFactory.HEADER_API_KEY to datadogContext.clientToken,
+            RequestFactory.HEADER_EVP_ORIGIN to datadogContext.source,
+            RequestFactory.HEADER_EVP_ORIGIN_VERSION to datadogContext.sdkVersion,
+            RequestFactory.HEADER_REQUEST_ID to requestId
+        )
+    }
+
+    private fun buildQueryParameters(datadogContext: DatadogContext): Map<String, Any> {
+        return mapOf(
+            QUERY_PARAM_SOURCE to datadogContext.source,
+            QUERY_PARAM_TAGS to tags(datadogContext)
         )
     }
+
+    private fun resolveRequest(context: DatadogContext, body: RequestBody?): Request {
+        if (body != null) {
+            val bodyAsByteArray = safeBodyToByteArray(body) ?: return getEmptyRequest()
+            val requestId = UUID.randomUUID().toString()
+            val description = "Session Replay Segment Upload Request"
+            val headers = resolveHeaders(context, requestId)
+            val requestUrl = buildUrl(context)
+            return Request(
+                requestId,
+                description,
+                requestUrl,
+                headers,
+                body = bodyAsByteArray,
+                contentType = body.contentType().toString()
+            )
+        } else {
+            return getEmptyRequest()
+        }
+    }
+
+    private fun safeBodyToByteArray(body: RequestBody): ByteArray? {
+        return try {
+            bodyToByteArrayFactory(body)
+        } catch (@Suppress("TooGenericExceptionCaught") e: Throwable) {
+            internalLogger.e(EXTRACT_BYTE_ARRAY_FROM_BODY_ERROR_MESSAGE, e)
+            null
+        }
+    }
+
+    // endregion
+
+    companion object {
+        internal const val EXTRACT_BYTE_ARRAY_FROM_BODY_ERROR_MESSAGE =
+            "Unable to extract the ByteArray from session replay request body."
+        internal const val QUERY_PARAM_SOURCE = "ddsource"
+        internal const val QUERY_PARAM_TAGS = "ddtags"
```

Comment:
minor: those exist in `RequestFactory` as well if needed.

Reply:
oh I didn't know that, good :)

Reply:
Do we add them by default there? Then I don't see the point for me to add them also here.

Reply:
no, there is only declaration there, they are not added automatically to any request.